Switch to packaging for parsing metadata and support metadata 2.4

dnicolodi opened the pull request:
The `packaging` package is maintained by the PyPA and it is the de-facto reference implementation for the packaging standards. Using `packaging` for parsing metadata guarantees support for the latest metadata versions.

`warehouse`, the Python package index implementation used by PyPI, also uses `packaging` for parsing metadata. This guarantees that metadata parsing is the same on the client and server side, for the most prominent index.

dnicolodi committed:
Switch from pkginfo to packaging for parsing distribution metadata

The packaging package is maintained by the PyPA and it is the de-facto
reference implementation for the packaging standards. Using packaging
for parsing metadata guarantees support for the latest metadata
versions.

warehouse, the Python package index implementation used by PyPI, also
uses packaging for parsing metadata. This guarantees that metadata
parsing is the same on the client and server side, for the most
prominent index.

dnicolodi committed:
Enable some more mypy checks

dnicolodi commented:
This passes the unit tests, except for this one https://github.com/pypa/twine/blob/dd613569856f1dd128040f197948b5c48eb8dd43/tests/test_package.py#L383-L437 which is not surprising as the metadata validation raises different exceptions now. I don't know whether the new code should be adapted to raise the same exceptions or it is acceptable to adjust the test.

dnicolodi commented:
I've adjusted the test mentioned above. The only remaining issue is that `packaging` and `pkginfo` and thus `twine` disagree on the name of two metadata fields: `pkginfo` uses `platform` and `supported_platform`, while `packaging` uses `platforms` and `supported_platforms`. I don't know who is right here.

dnicolodi committed:
Move monkeypatching of metadata 2.0 support to a more proper place

It was done in the support code for the wheel file format but it
affects metadata loading from all supported distribution types. Move
it to generic code.

dnicolodi committed:
Accommodate for invalid metadata produced by setuptools

See pypa/setuptools#4759.

konn commented:
Is there any plan to merge this anytime soon or support metadata 2.4 in other ways?

Our product uses `maturin` as the build system and its new version 1.7.6 released today uses metadata version 2.4, and hence wheel generated by it doesn't check with the currently released version of twine 5.1.1.

woodruffw commented:
> Is there any plan to merge this anytime soon or support metadata 2.4 in other ways?

Once #1123 lands in a release, metadata 2.4 will be supported whenever a newer (> 1.10) version of `pkginfo` is resolved. This isn't fully general unfortunately, however, since existing `twine` installations that are `pip upgrade`'d won't get their deps upgraded at the same time. This is mentioned in #1070.

I would (personally) like to merge this soon as a fully general solution to the above. 

To whit: @dnicolodi would you be able to deconflict here? Once this is deconflicted I can do a full review pass.

dnicolodi commented:
> Once #1123 lands in a release, metadata 2.4 will be supported whenever a newer (> 1.10) version of `pkginfo` is resolved.

This is not entirely true: `twine` will not error out anymore on metadata version 2.4, but it will not transmit the new license fields to the package index. In the case of PyPI this will result in the license information not shown on the package page on PyPI. As the goal of PEP 639 and metadata version 2.4 is to make licensing information more clear, not making the license information available from the package index IMHO does not count as supporting metadata 2.4.

Furthermore, metadata 2.4 strongly discourages keeping license information in the package classifiers. If package authors follow this recommendation and upload to PyPI with twine their packages will not snow any licensing information on PyPI. This would be a considered a regression by all the package authors that were considering to be early adopters of metadata 2.4 (or that are forced to metadata 2.4 by build backend authors switching their backends to emit metadata 2.4 unconditionally, `maturin` and `pdm` being two widely used examples).

dnicolodi commented:
Rebased, twice.

jaraco commented on tests/test_wheel.py:
```diff
@@ -32,8 +32,7 @@
 def example_wheel(request):
-    file_name = os.path.join(helpers.TESTS_DIR, request.param)
-    return wheel.Wheel(file_name)
+    return wheel.Wheel(os.path.join(helpers.TESTS_DIR, request.param))
```
Nitpick: This change appears to me to be a no-op stylistic change. I guess it's fine, but also feels out of scope for this PR.

jaraco commented on twine/package.py:
```diff
@@ -18,39 +18,35 @@
     "bdist_wininst": wininst.WinInst,
-    "bdist_egg": pkginfo.BDist,
-    "sdist": pkginfo.SDist,
+    "bdist_egg": bdist.BDist,
+    "sdist": sdist.SDist,
```
I'm uneasy replacing an established implementation with an inline implementation of something very similar. Can you point to the rationale for re-implementing parts of pkginfo? My preference is to converge implementations and not fork them.

dnicolodi replied:
I can maybe use the implementation in `pkginfo` but to me it seems unnecessary to have a dependency to an external package to extract a member from a zip or tar archive. `twine` already implements it for two distribution types. Also, I have the impression that the `pkginfo.Distribution.read()` method is an implementation detail.

jaraco commented on twine/package.py:
```diff
@@ -133,104 +133,51 @@ def from_filename(cls, filename: str, comment: Optional[str]) -> "PackageFile":
-                    "Make sure the distribution includes the files where those fields "
-                    "are specified, and is using a supported Metadata-Version: "
-                    f"{', '.join(supported_metadata)}."
+        # setuptools emits License-File metadata fields while declaring
+        # Metadata-Version 2.1. This is invalid because the metadata
+        # specification does not allow to add arbitrary fields, and because
+        # the semantic implemented by setuptools is different than the one
+        # described in PEP 639. However, rejecting these packages would be
+        # too disruptive. Drop License-File metadata entries from the data
+        # sent to the package index if the declared metadata version is less
+        # than 2.4.
```
Thanks for the explanation here. I'm worried this change could also be disruptive. What's the harm in allowing the metadata fields to pass as they had before? Is it necessary to support metadata 2.4 (this issue)?

dnicolodi replied:
They are refused by PyPI.

jaraco commented on twine/package.py:
```diff
@@ -133,104 +133,51 @@ def from_filename(cls, filename: str, comment: Optional[str]) -> "PackageFile":
-        # which could also be empty if, for example, a MANIFEST.in doesn't include
-        # setup.cfg.
-        missing_fields = [
-            f.capitalize() for f in ["name", "version"] if not getattr(meta, f)
```
Here's where I see we lost capitalization (and canonical names) of the fields missing in the message.

dnicolodi replied:
Not exactly. The change in error messages are a consequence of using `packaging.metadata` for metadata parsing and validation.

jaraco commented on twine/repository.py:
```diff
@@ -25,7 +25,14 @@
-KEYWORDS_TO_NOT_FLATTEN = {"gpg_signature", "attestations", "content"}
+KEYWORDS_TO_NOT_FLATTEN = {"gpg_signature", "attestations", "content", "keywords"}
+
+# Map ``packaging.metadata.RawMetadata`` fields to package upload form fields.
```
Is it not possible to simply use the same singular or plural keyword to match? Why are `classifiers` not affected by this need to rename?

dnicolodi replied:
The field names in `packaging.metadata.RawMetadata` are not under the control of `twine` and using `packaging.metadata.RawMetadata` is implicit in using `packaging.metadata` for parsing and validating metadata. Somewhere the rename needs to happen. Renaming here makes the code easier everywhere else. The `packaging` maintainers have reasons for naming the fields like this, see the `packaging` documentation for their reasoning.

jaraco commented on twine/repository.py:
```diff
@@ -91,12 +98,21 @@ def close(self) -> None:
+            key = KEYWORDS_TO_RENAME.get(key, key)
+            if isinstance(value, (list, tuple)) and key not in KEYWORDS_TO_NOT_FLATTEN:
                 data_to_send.extend((key, item) for item in value)
+            elif key == "project_urls":
+                assert isinstance(value, dict), "for mypy"
+                for name, url in value:
+                    data_to_send.append((key, f"{name}, {url}"))
+            elif key == "keywords":
+                assert isinstance(value, list), "for mypy"
+                data_to_send.append((key, ", ".join(value)))
+            else:
+                data_to_send.append((key, value))
```
I'm concerned about how this code replaces code that was previously largely agnostic to the specific keys. Now it contains two special cases. Can we refactor this so it behaved more or less how it did before (i.e. either flatten or don't)... or maybe let each of the data.items provide a rendering function? I'd like to avoid entangling concerns, and this logic has a lot of entanglement. For example, `keywords` is reached because it's not in KEYWORDS_TO_NOT_FLATTEN, but `project_urls` is reached because it's not a list or tuple. I'd like to get back to an implementation that's less complicated.

dnicolodi replied:
The code already contained special case in the handling of a few keys. It was just hiding the special keys into a module global variable. Now there are more special cases, but not that many more. The names of the metadata fields in `pyproject.toml`, in `METADATA` and in the submission form are not consistent. Thus there needs to be a rename somewhere. Then there are fields that are represented differently in the form, and there are two cases in which `packaging.metadata.parse_email()` decomposes a structured field into its components for validation and the code here needs to put them back into textual representation.

Changing the code to have all the values in `PackageMetadata` to be object with a serialization method seems much more complex that what this function does. The logic to implement would still need to be the same, but in that implementation it would be diluted into boilerplate code. But maybe I'm misunderstanding what you are suggesting.

I can change the order of the conditionals to have `keywords` handled first, so that the logic is a bit simpler.

dnicolodi replied:
I've restructured the code. It is not easier to read and it is evident that only two special cases are added to the one existing.

jaraco commented on twine/repository.py:
```diff
@@ -91,12 +98,21 @@ def close(self) -> None:
+            key = KEYWORDS_TO_RENAME.get(key, key)
+            if isinstance(value, (list, tuple)) and key not in KEYWORDS_TO_NOT_FLATTEN:
                 data_to_send.extend((key, item) for item in value)
+            elif key == "project_urls":
+                assert isinstance(value, dict), "for mypy"
+                for name, url in value:
+                    data_to_send.append((key, f"{name}, {url}"))
```
I'm confused here. Is `value` a `dict` with keys that are `(name, url)` pairs?

I removed this elif block, and the tests still pass. If this code is necessary, it needs some test coverage.

dnicolodi replied:
`twine.package.PackageMetadata` is a typed dictionary, if you look at its type annotation you will see that indeed `project_urls` is a dictionary of `name` and `url` pairs. I see that the code contains a bug, thus tests are needed, there are no existing tests for this function.

dnicolodi replied:
I added a test.

jaraco commented on twine/repository.py:
```diff
@@ -108,7 +124,7 @@ def set_client_certificate(self, clientcert: Optional[str]) -> None:
 
     def register(self, package: package_file.PackageFile) -> requests.Response:
-        data = package.metadata_dictionary()
+        data = dict(**package.metadata_dictionary())
```
I reverted this change and the tests still pass. Why `dict(**)`? It would be better if the result from `package.metadata_dictionary()` did not require the caller to wrap the result in `dict(**)` or `dict()`. How can that be achieved?

dnicolodi replied:
The `**` are indeed not necessary. They are a remnant of a previous approach. The `dict()` is just to make mypy happy: the code adds keys to the dictionary outside the typed keys and mypy does not like that. I can replace `dict()` with a cast if you prefer.

dnicolodi replied:
I restructured the code to do not require the type casting anymore.

jaraco commented on twine/wheel.py:
```diff
@@ -35,27 +27,22 @@
 )
 
 
+def find_candidate_metadata_files(names: List[str]) -> List[List[str]]:
+    """Filter files that may be METADATA files."""
+    tuples = [x.split("/") for x in names if "METADATA" in x]
+    return [x[1] for x in sorted((len(x), x) for x in tuples)]
```
nitpick: making these stylistic changes should be reserved for a separate PR.

dnicolodi replied:
Dropped this change.

jaraco reviewed (changes requested):
Overall, this is pretty good.

I have a few comments above, mostly with a theme of "can we make this change less invasive?" I imagine this change could have been broken down into support for metadata 2.4, switching from pkginfo to packaging, and other unrelated changes.

Nitpicks are my preference, but not required.

Additionally:

- [x] Add a changelog.

dnicolodi commented:
> I imagine this change could have been broken down into support for metadata 2.4, switching from pkginfo to packaging, and other unrelated changes.

The changes are broken down into independent commits with the reason for each change explained in the commit message. I am unfamiliar with the development practices of this package. If the practice here is to squash the commits in each PR into a single commit, I can split the commits into separate PRs, if you prefer.

dnicolodi commented on pyproject.toml:
```diff
@@ -69,7 +68,7 @@ twine = "twine.__main__:main"
 packages = ["twine", "twine.commands"]
 include-package-data = true
-license-files = ["LICENSE"]
+license-files = []
```
Without this change, without the last commit in the PR, twine fails to validate its own metadata or to upload itself. With the last commit in the PR this is not an issue anymore, but I thought that emitting correct metadata would be desirable anyway. I can drop this commit, if you prefer.

dnicolodi replied:
I dropped this commit.

dnicolodi commented on tests/test_package.py:
```diff
@@ -31,7 +31,7 @@ def test_sign_file(monkeypatch):
         filename=filename,
         comment=None,
-        metadata=pretend.stub(name="deprecated-pypirc"),
+        metadata=dict(name="deprecated-pypirc", version="1.2.3"),
```
The `metadata` argument to the `PackageFile` constructor is not anymore the object produced by `pkginfo` (I'm too lazy to check how it is called) but a `packaging.metadata.RawMetadata` typed dict. `version` is a required metadata entry, thus the code in the `PackageFile` constructor assumes it is there. These changes simply reflect that.

sigmavirus24 replied:
@dnicolodi in that case, it might be better to explicitly use that dict class instead of just `dict(...)`

```suggestion
        metadata=packaging.metadata.RawMetadata(name="deprecated-pypirc", version="1.2.3"),
```

dnicolodi replied:
Suggested changes applied.

dnicolodi commented on tests/test_package.py:
```diff
@@ -172,86 +172,90 @@ def test_package_safe_name_is_correct(pkg_name, expected_name):
 
 
 def test_metadata_dictionary_keys():
```
I wanted to keep the changes to a minimum. In spirit this test does exactly the same thing: verify the fields in the dictionary returned by the `metadata_dictionary()` method, thus the test name didn't see too far off. Anyway, I'll rename.

dnicolodi commented on tests/test_package.py:
```diff
@@ -172,86 +172,90 @@ def test_package_safe_name_is_correct(pkg_name, expected_name):
-        keywords=pretend.stub(),
-        platforms=pretend.stub(),
-        classifiers=pretend.stub(),
+        supported_platform=pretend.stub(),
```
It is how the field in called in `packaging.metadata.RawMetadata`. The code is switched to use that as as base for storing metadata, thus the rename.

dnicolodi commented on tests/test_package.py:
```diff
@@ -381,76 +387,64 @@ def test_fips_metadata_excludes_md5_and_blake2(monkeypatch):
     with pytest.raises(exceptions.InvalidDistribution) as err:
         package_file.PackageFile.from_filename(filename, comment=None)
 
-    assert (
-        f"Metadata is missing required fields: {missing_fields}." in err.value.args[0]
-    )
```
The exception message is simply a concatenation of the validation errors returned by `packaging.metadata.Metadata.from_raw()`. I can change how they are concatenated, but it is unrealistic to change them otherwise.

dnicolodi committed:
Remove "content" from set of specially handled metadata fields

The "content" field is always added to the form data after the package
metadata has been flattened, thus it is not needed to handle it in the
flattening method. Remove the associated test.

This will allow to tighten typing in a successive commit.

dnicolodi committed:
Remove "attestations" from the set of specially handled metadata fields

The "attestations" field is a string: strings do not need flattening.

dnicolodi committed:
Refactor code a tiny bit

Avoid looking a key up into a set of one element and remove an
indirection through a module global variable. This will make it a bit
easier to extend the flattening logic in successive commits.

dnicolodi commented:
Thanks @jaraco for the review. I addressed some of your comments and replied to the remaining. Please have another look.

> I imagine this change could have been broken down into support for metadata 2.4, switching from pkginfo to packaging, and other unrelated changes.

There is exactly one line of code in this PR that introduces support for metadata 2.4. The remaining is preparation to switch from pkginfo to packaging, the switch itself, some minor cleanup, and a fix to allow to upload distributions with invalid metadata created by setuptools. I split the changes in what I think are easy to review commits. I don't think that splitting the commits in multiple PRs makes the review any easier, but I can do that if it helps.

satmandu commented:
Thanks for working on this. Hopefully this will also address uploading wheels of ruff 0.8.1 which as discussed in https://github.com/astral-sh/ruff/issues/14681 uses metadata 2.4.

dnicolodi commented:
@satmandu The issue you link to is about `poetry` not supporting metadata 2.4 and `maturin` (the build backend used by `ruff`) having switched to produce metadata 2.4 by default. `ruff` uses `uv publish` to upload its distributions to PyPI https://github.com/astral-sh/ruff/blob/76d2e565019b1561181595da4acd5969187f1179/.github/workflows/publish-pypi.yml#L32 `twine` is not involved at all.

sigmavirus24 commented on tests/test_package.py:
```diff
@@ -51,7 +51,7 @@ def test_sign_file_with_identity(monkeypatch):
         filename=filename,
         comment=None,
-        metadata=pretend.stub(name="deprecated-pypirc"),
+        metadata=dict(name="deprecated-pypirc", version="1.2.3"),
```
```suggestion
        metadata=packaging.metadata.RawMetadata(name="deprecated-pypirc", version="1.2.3"),
```

sigmavirus24 commented on tests/test_package.py:
```diff
@@ -106,7 +106,7 @@ def test_package_signed_name_is_correct():
         filename=filename,
         comment=None,
-        metadata=pretend.stub(name="deprecated-pypirc"),
+        metadata=dict(name="deprecated-pypirc", version="1.2.3"),
```
```suggestion
        metadata=packaging.metadata.RawMetadata(name="deprecated-pypirc", version="1.2.3"),
```

sigmavirus24 commented on tests/test_package.py:
```diff
@@ -163,95 +163,100 @@ def test_package_safe_name_is_correct(pkg_name, expected_name):
         filename="tests/fixtures/deprecated-pypirc",
         comment=None,
-        metadata=pretend.stub(name=pkg_name),
+        metadata=dict(name=pkg_name, version="1.2.3"),
```
```suggestion
        metadata=packaging.metadata.RawMetadata(name=pkg_name, version="1.2.3"),
```

sigmavirus24 commented on tests/test_package.py:
```diff
@@ -163,95 +163,100 @@ def test_package_safe_name_is_correct(pkg_name, expected_name):
-        # identify release
+def test_supported_metadata_fields():
+    """Check that ``twine.package.PackageMetadata`` contains the expected fields."""
+    assert set(package_file.PackageMetadata.__annotations__) == {
```
Relying on `__annotations__` breaks what is being tested here. The test here is that we serialize this to a dictionary for sending through the HTTP client as a `multipart/form-data` upload. Not that a class has the right fields.

dnicolodi replied:
My reading of the test is that it just checked the dictionary fields. The serialization to `multipart/form-data` happens at a different layer and before that the metadata goes through transformation in `twine.repository.Repository._convert_data_to_list_of_tuples()` (now `twine.repository.Repository._convert_metadata_to_list_of_tuples()`). This just checked that a dictionary with the right keys is instantiated. It relied on dictionary entries were inserted with `None` value for metadata fields without a value (which would be invalid if `warehouse` were not to have explicit support for it to accommodate `twine`, BTW). Now fields that are not present in the package metadata are not populated and `packaging` is responsible for metadata validation. I decided to keep the test to verify that `packaging` does not rename, removes, or add dictionary entries.

I added a test in `test_repository.py` that tests something closer to the `multipart/form-data` format, but it ides not check the filed names. If we think that checking the field names is important, I can add it there.

dnicolodi replied:
Also, the `test_metadata_dictionary_values()` also tests the metadata fields, but for how it is structured, it cannot test for the presence of fields unknown to `twine`. I think the two tests together cover all is relevant to test.

sigmavirus24 replied:
@dnicolodi so to be clear, this isn't testing the `multipart/form-data` encoding. This is testing that the fields we know are valid and are passed through, are placed into the dictionary that is then used to generate the `multipart/form-data` encoding. Again, `__annotations__` is not the same as testing the supported metadata fields.

To be clear, there are different things here:

1. You're testing a TypedDict here that is inheriting from a dict we do not control. We don't run tests frequently enough to detect a problem with `packaging` introduced somewhere before `packaging` or `twine` releases so we risk breaking others by having a non-pinned `packaging` version.
2. The libraries we're using (and which most other libraries mimic) are then expecting  a dictionary. We don't need to test those libraries and that they produce the proper serialized data, but instead that what we pass to them is consistent. That was the purpose of this test. I _know_ because _I_ wrote it.
3. You're test in `test_repository` isn't testing `multipart/form-data` encoding either. It's testing how we prefer to pass it into there such that items are vaguely ordered. You're also effectively testing a private method which doesn't seem valuable it just makes it harder to refactor.

sigmavirus24 replied:
You're missing the point and either GitHub lost a comment with your changes or something else happened (where it didn't properly post) but our PackageMetadata object should _not_ inherit from packaging's RawMetadata class. That's a very fragile assumption to make about the stability/implementation of that class. If this were done well, we'd use it as a collaborator, not as a parent. It's not a parent, we don't own it. As a result, your test change here doesn't make sense.

dnicolodi replied:
> If this were done well ... your test change here doesn't make sense

I would like you to use a more friendly tone. I am not your subordinate.

sigmavirus24 commented on tests/test_package.py:
```diff
@@ -163,95 +163,100 @@ def test_package_safe_name_is_correct(pkg_name, expected_name):
-    meta = pretend.stub(
-        name="whatever",
-        version=pretend.stub(),
+    meta = dict(
```
```suggestion
    meta = packaging.metadata.RawMetadata(
```

sigmavirus24 commented on twine/package.py:
```diff
@@ -18,39 +18,35 @@
-
-import packaging.version
-import pkginfo
+from typing import Any, Dict, List, NamedTuple, Optional, Tuple
```
As much as possible (where you're modifying something like this) please split these into separate lines. That avoids merge conflicts. You already did it L23-L24, just be consistent on the better way of formatting

dnicolodi replied:
Do you mean one `import` statement per symbol imported (as in L23-24) or with parenthesized list of of symbols as in the lines replaced here?

sigmavirus24 replied:
Either is fine, but the former is what I was advocating for.

It's also already the pattern below with `from twine import _` and `from packaging import _`.

dnicolodi replied:
It is a pattern never used in the codebase when importing from `typing`. I wrote it like this for consistency, but if you prefer one import per line, I can change it.

dnicolodi replied:
The style I used (all symbols on one line) is actually enforced by the configuration of `isort` used for the project. What should I do?

sigmavirus24 reviewed (changes requested):
Still haven't finished goin gthrough everything, first set of comments though

di commented on twine/package.py:
```diff
@@ -18,39 +18,35 @@
 from twine import wheel
 from twine import wininst
 
+# Monkeypatch Metadata 2.0 support
+metadata._VALID_METADATA_VERSIONS = [
+    "1.0",
+    "1.1",
+    "1.2",
+    "2.0",
+    "2.1",
+    "2.2",
+    "2.3",
+    "2.4",
+]
+
```
Note that PyPI hasn't supported Metadata 2.0 since March of this year, and we don't plan to add it back, so this may be unnecessary: https://github.com/pypi/warehouse/issues/17168

dnicolodi replied:
I know, I think I'm the one that noticed it first, see pypi/warehouse#17168 😄. However this PR does already many things. I would leave dropping support for metadata 2.0 to another conversation.

sigmavirus24 replied:
As a reminder, twine supports more than just PyPI

di replied:
@dnicolodi Aha! Noted, and thanks for catching that :)

sigmavirus24 replied:
Agreed @dnicolodi let's keep this PR as small as possible. But again, there are other package indexes that people use twine with that aren't PyPI. We have no way of knowing what they expect/people are expecting to use with them.

wimglenn replied:
This monkeypatch breaks twine for Metadata-Version 2.5. Recently the `metadata._VALID_METADATA_VERSIONS` in packaging includes "2.5" ([src](https://github.com/pypa/packaging/blob/26.2/src/packaging/metadata.py#L514)) but the monkeypatch overwrites it. Took me ages to figure out what spooky action at a distance was going on!  🥴

dnicolodi commented:
> Still haven't finished goin gthrough everything, first set of comments though

Thanks for the review @sigmavirus24. I addressed your comments, except for one for which I have a question and one for which I have a rebuttal.

sigmavirus24 commented on docs/conf.py:
```diff
@@ -291,7 +292,6 @@
 # TODO: Try to add these to intersphinx_mapping
 nitpick_ignore_regex = [
     (r"py:.*", r"pkginfo.*"),
```
Do we need `pkginfo` anymore here?

sigmavirus24 commented on twine/package.py:
```diff
@@ -78,19 +72,23 @@ def _safe_name(name: str) -> str:
-
-    name: str
-    version: str
+class PackageMetadata(metadata.RawMetadata, total=False):
```
I'm not a fan of sub-classing here. 

This opens up an issue where `metadata.RawMetadata` introduces some other private field and that then leads to a potential issue with serializing it properly.

Instead, an object (dataclass - as poor as they may be for performance, a different TypedDict with the `metadata.RawMetadata` as an attribute, a plain old typed python object) with this that can do the known serialization properly is better. This relates to the test you changed in `test_package.py`.

dnicolodi replied:
I absolutely agree with the concern: `twine` should not use metadata fields it does not know about.

However, `packaging.metadata.RawMetadata` is public API thus it cannot gain private fields: it is supposed to be a simple representation of the serialized (as in written in RFC822 format in the `METADATA` or `PKG-INFO` files) metadata (see the `packaging` [documentation](https://packaging.pypa.io/en/stable/metadata.html#packaging.metadata.RawMetadata)). Therefore, restricting the supported metadata versions to the currently supported metadata versions (see lines 33 to 43 in `package.py`) ensures that only the metadata fields for these metadata versions are populated in `packaging.metadata.RawMetadata`. 

Indeed the test in `test_package.py` would break for a future `packaging` release that implements support for a new metadata version that adds some more fields, but that is a problem of the test, not of the implementation (the fields in `metadata.RawMetadata` cannot be populated thus they would not be emitted in serialization).

I understand that this "contractual" limit to the populated metadata fields feels weaker than something enforced in code, however, there is the same likelihood of `packaging` breaking this promise as it breaking the representation of the fields or their meaning, and that cannot be independently checked without reimplementing metadata parsing and validation.

A possible objection to this approach is that `twine` should validate correctly distributions using metadata versions it does not know about and upload only the subset of the metadata to the package index. However, this approach would leave the possibility open for mishandling metadata fields that may change meaning or format in future metadata versions. And partial support of a metadata version is not that great of an experience for the users. On the other hand, new metadata versions are not published so frequently to make verifying support is `twine` problematic.

sigmavirus24 replied:
> However, packaging.metadata.RawMetadata is public API thus it cannot gain private fields

This statement makes little to no sense to me. It's a public API, and there are ways to keep that backwards compatibility while adding in private fields and annotations for them.

The `__annotation__` on that object are _not_ a public API though.

dnicolodi replied:
> This statement makes little to no sense to me. It's a public API, and there are ways to keep that backwards compatibility while adding in private fields and annotations for them.

It is a `TypedDict` what kind of private fields can be added to it? You decided to gloss over the API stability promises provided by the `packaging` developers. I think you are too vested into declaring my code unfit to give proper consideration to my argument. However, I don't care enough to debate this further. I'll change the code.

> The `__annotation__` on that object are _not_ a public API though

I don't understand what you mean. `__annotations__` is public API https://docs.python.org/3/howto/annotations.html#annotations-howto

> Really? Haven't there been at least 3 versions in just the last couple years?

- Metadata 1.0 - PEP 241, 19-Mar-2001
- Metadata 1.1 - PEP 314, 29-Apr-2003
- Metadata 1.2 - PEP 345, 22-Dec-2009
- Metadata 2.1 - PEP 566, 01-Dec-2017
- Metadata 2.2 - PEP 643, 14-Nov-2020
- Metadata 2.3 - PEP 685, 08-Mar-2022 (no metadata fields added or removed)
- Metadata 2.4 - PEP 639, 10-May-2024

These are the PEP posting dates as PEP acceptance dates are not recorded. Taking "a couple of years" to mean the customary two years, we have had exactly one metadata version published. Extending it to mean four years, we still only have on metadata version published that introduced new metadata fields. To have 3 metadata versions adding fields we need to go back 7 years. I don't want to debate whether "a couple" can be stretched to mean 7. I just care about the implications of your assertion.

Are you suggesting that `twine` should not limit the metadata versions he accepts to the ones it knows about and instead leave the upper bound open ended?

sigmavirus24 commented on tests/test_repository.py:
```diff
@@ -37,37 +55,49 @@ def test_gpg_signature_structure_is_preserved():
     assert tuples == [("platform", "UNKNOWN"), ("platform", "ANOTHERPLATFORM")]
 
 
+def test_form_metadata_translation(example_distribution):
```
Again, you're not testing "translation" or serialization here, you're testing a private method which is already well tested above. I don't believe this test adds any real value.

dnicolodi replied:
I removed this test.

sigmavirus24 commented on tests/test_repository.py:
```diff
@@ -37,37 +55,49 @@ def test_gpg_signature_structure_is_preserved():
+        assert isinstance(value, str) or value is None
+
+
+def test_form_metadata_translation_everything(monkeypatch):
```
And this is a variation on the above test which again, doesn't seem as helpful.

dnicolodi replied:
I'll remove this too.

dnicolodi replied:
I think there is value in keeping this test. I added some comments explaining what the test does. It should be clearer now.

sigmavirus24 commented on twine/bdist.py:
```diff
@@ -0,0 +1,61 @@
+# Copyright 2024 Daniele Nicolodi
```
My understanding is that much of what is here you cribbed from other libraries with their own copyright holders and licensing, thus you are not allowed to claim copyright or re-license it.

dnicolodi replied:
I think there is value in keeping this test. I added some comments explaining what the test does. It should be cleared now.

sigmavirus24 commented on twine/sdist.py:
```diff
@@ -0,0 +1,40 @@
+# Copyright 2024 Daniele Nicolodi
```
Same here.

Also, I believe the old copyright notices are:
A. Unnecessary
B. Should probably get removed at some point

dnicolodi replied:
Same reply as above. I just followed the pattern in all other source files in `twine`.

sigmavirus24 commented on twine/wininst.py:
```diff
@@ -3,21 +3,16 @@
 
     @property
-    def py_version(self) -> str:
+    def py_version(self) -> Optional[str]:
```
Everything at this level of `twine` is intended and in-scope for another project to import. Changing the signatures like this is thus a backwards incompatible change.

Are folks actually importing these? Probably not. But they could be and that's a risk I'm worried about. We'll need to make sure we release this as `7.0`.

dnicolodi replied:
This is `Optional[str]` to keep the interface unified with all the implementations of `twine.distribution.Distribution`. The only subclass that returns `None` is however `twine.sdist.SDist` and it should rather return `"source"` as indicated in the documentation for the package index submission format. However, I didn't wan to add another cleanup on top. I can submit another PR to do that and then restore this to return `str`.

dnicolodi replied:
I restored the signature. See also #1191.

sigmavirus24 reviewed (changes requested):
Probably need to take another pass at some point

dnicolodi commented:
Rebased on top of the recent changes.

dnicolodi commented:
Updated. Please take a look.

jaraco commented:
> > I imagine this change could have been broken down into support for metadata 2.4, switching from pkginfo to packaging, and other unrelated changes.
> 
> The changes are broken down into independent commits with the reason for each change explained in the commit message. I am unfamiliar with the development practices of this package. If the practice here is to squash the commits in each PR into a single commit, I can split the commits into separate PRs, if you prefer.

That's fair. Sorry if I was too nitpicky. I didn't think to read the history of commits, but I probably could have for more context.

jaraco commented on changelog/1180.misc.txt:
```diff
@@ -0,0 +1,8 @@
+- ``packaging`` is used instead of ``pkginfo`` for parsing and
+  validating metadata. This aligns metadata validation to the one
+  performed by PyPI. ``pkginfo`` is not a dependency anymore.
+- Metadata fields defined added with metadata version 2.4 as defined
+  by PEP 639 are now sent to the package index when a distribution is
+  uploaded. This results in licensing information to appear correctly
+  on the package page on PyPI when uploading packages using metadata
+  version 2.4.
```
I'm not sure if this format works or if it'll need to be reformatted. In other projects, when I've had the need to describe two changes, I create two changelog entries (both with the same issue number).

jaraco commented on tests/test_package.py:
```diff
@@ -381,76 +387,64 @@ def test_fips_metadata_excludes_md5_and_blake2(monkeypatch):
+        ),
+        pytest.param(
+            b"Metadata-Version: 2.3\n",
+            "'name' is a required field, 'version' is a required field",
```
Shame that `packaging` doesn't use the canonical form of the proper names.

woodruffw commented on twine/distribution.py:
```diff
@@ -0,0 +1,8 @@
+class Distribution:
```
Nitpick, nonblocking: maybe this should be an ABC or an interface, assuming it isn't intended to be instantiated directly?

dnicolodi replied:
What do you mean with interface? A `typing.Protocol`? Protocols are useful when direct inheritance is impossible or not convenient to implement. Here we have control of all subclasses, thus I don't see a protocol adding any value. An ABC would be more appropriate, but for a simple case like this I see it as adding more noise than value. I can use an ABC if there strong desire for it.

woodruffw reviewed (approved):
LGTM, thank you for your hard work here @dnicolodi! One non-blocking comment.

DanielNoord commented on changelog/1180.misc.txt:
```diff
@@ -0,0 +1,8 @@
+- ``packaging`` is used instead of ``pkginfo`` for parsing and
+  validating metadata. This aligns metadata validation to the one
+  performed by PyPI. ``pkginfo`` is not a dependency anymore.
+- Metadata fields defined added with metadata version 2.4 as defined
```
```suggestion
- Metadata fields added with metadata version 2.4 as defined
```

dnicolodi replied:
Thanks!

DanielNoord commented on pyproject.toml:
```diff
@@ -41,8 +40,8 @@ dependencies = [
 	"rich >= 12.0.0",
-	"packaging",
-	"id",
+	"packaging >= 24.0",
```
Please correct me if I'm wrong but support for PEP 639 (as mentioned in the changelog entry) was only added in 24.2. See release notes: https://github.com/pypa/packaging/releases/tag/24.2

Shouldn't this be limited to `>= 24.2`?

dnicolodi replied:
It is correct that support for metadata 2.4 ha been added to `packaging` only in version 24.2. However, using `packaging` to parse metadata only requires functionality added in version 24.0. With `packaging` version < 24.2 >= 24.0 everything added here still works, thus I didn't feel like requiring the latest version.

DanielNoord replied:
Makes sense, but I think the changelog should reflect this. From reading the Discourse thread this is eagerly awaited. Making it so that `pip install -U twine` doesn't actually include support for `2.4` but that `pip install -U packaging` is also needed will probably confuse people.

DanielNoord reviewed (approved):
Maintainers can probably comment on whether they think this too verbose, but I think this explains it very well.

Only looked at the project metadata and changelog. From outside perspective and an interested passer-by that LGTM! Thanks for picking this up on behalf of many within the ecosystem waiting on this :)

sigmavirus24 commented on twine/package.py:
```diff
@@ -72,29 +65,120 @@ def _safe_name(name: str) -> str:
+    license_file: List[str]
+
+    # Additional metadata
+    comment: Optional[str]
+    pyversion: str
+    filetype: str
+    gpg_signature: Tuple[str, bytes]
+    attestations: str
+    md5_digest: str
+    sha256_digest: Optional[str]
+    blake2_256_digest: str
```
These aren't actually metadata in the way the rest of these fields are. These are internal-ish data about the package file in question. In other words, there's a real and valuable distinction between "I parsed this from the package metadata" and "I generated this from the package file to be uploaded and/or its metadata". Blurring those lines may be convenient-ish, but long term will be confusing for new contributors, maintainers, etc. I don't believe these should live here. As I said, we can use packaging's typed dict as a collaborator here, but I won't agree to inheriting from it or doing something like this

dnicolodi replied:
Thanks @sigmavirus24 for the review. However, I think there is a misunderstanding.

The type definition you are commenting on is defining the return type of the `Package.metadata_dictionary()`: https://github.com/pypa/twine/blob/1703ae7316f73358abcc01c2cf0df6b922094a3f/twine/package.py#L178-L182

As you can see, the goal of the function is indeed to merge multiple sources of "metadata". This is existing code. The only thing that my patch does it to type the returned value as a TypedDict. I thought that using a TypedDict made things clearer and resolved some ambiguities regarding the data types stored in the dictionary. If adding typing is more confusing than helpful, I can remove it. I agree that the naming of the method is not the most clear: what is returned is the package submission data, not the package metadata, but, again, that is the existing name of the method. I named the TypedDict to match the method name. I can change that to something else, if we can find a more fitting name.

> As I said, we can use packaging's typed dict as a collaborator here, but I won't agree to inheriting from it or doing something like this

Things are already structured like this: the `Package` class has a `metadata` member that holds the parsed metadata as returned by `packaging` and the additional fields required for submission: https://github.com/pypa/twine/blob/1703ae7316f73358abcc01c2cf0df6b922094a3f/twine/package.py#L82-L96
or in the proposed patch:
```python
class PackageFile:
    def __init__(
        self,
        filename: str,
        comment: Optional[str],
        metadata: metadata.RawMetadata,
        python_version: str,
        filetype: str,
    ) -> None:
        self.filename = filename
        self.basefilename = os.path.basename(filename)
        self.comment = comment
        self.metadata = metadata
        self.python_version = python_version
        self.filetype = filetype
```
However, at some point the different sources of data for the submission form need to be combined into a single data structure. The existing code uses a dictionary for that.

I hope that this clarification resolves the issue you have with the code structure. If not, I may be misunderstanding your comment. In this case I would like to point out that all the proposed patch does is to change the source of the metadata values. The combining into a single dictionary of all the value required for submission is existing code. If you prefer to structure this code differently, maybe you can change the existing code to a structure you like more, and I'll rebase my changes on top.

sigmavirus24 replied:
Ah that does clarify things. Sorry for the confusion. I saw our new maintainers approval and could only review from my phone which makes it harder to follow some things. Sorry for the confusion.

dnicolodi replied:
> Sorry for the confusion.

No problem. Thank you for the review and for merging!
